java/gandiva/CMakeLists.txt

```diff
@@ -23,12 +23,41 @@ include(UseJava)
 include(FindJNI)
 
 message("generating headers to ${JNI_HEADERS_DIR}/jni")
-add_jar(gandiva_java
-        src/main/java/org/apache/arrow/gandiva/evaluator/ConfigurationBuilder.java
-        src/main/java/org/apache/arrow/gandiva/evaluator/JniWrapper.java
-        src/main/java/org/apache/arrow/gandiva/evaluator/ExpressionRegistryJniHelper.java
-        src/main/java/org/apache/arrow/gandiva/exceptions/GandivaException.java
-        GENERATE_NATIVE_HEADERS
-        gandivajni_headers
-        DESTINATION
-        ${JNI_HEADERS_DIR}/jni)
+
+# generate_native_headers is available only from java8
+# centos5 does not have java8 images, so supporting java 7 too.
+# unfortunately create_javah does not work in java8 correctly.
+if(ARROW_GANDIVA_JAVA7)
+  add_jar(
+    gandiva_java
+    src/main/java/org/apache/arrow/gandiva/evaluator/ConfigurationBuilder.java
+    src/main/java/org/apache/arrow/gandiva/evaluator/JniWrapper.java
+    src/main/java/org/apache/arrow/gandiva/evaluator/ExpressionRegistryJniHelper.java
+    src/main/java/org/apache/arrow/gandiva/exceptions/GandivaException.java)
+
+  create_javah(TARGET
+               gandiva_jni_headers
+               CLASSES
+               org.apache.arrow.gandiva.evaluator.ConfigurationBuilder
+               org.apache.arrow.gandiva.evaluator.JniWrapper
+               org.apache.arrow.gandiva.evaluator.ExpressionRegistryJniHelper
+               org.apache.arrow.gandiva.exceptions.GandivaException
+               DEPENDS
+               gandiva_java
+               CLASSPATH
+               gandiva_java
+               OUTPUT_DIR
+               ${JNI_HEADERS_DIR}/jni)
+else()
+  add_jar(
+    gandiva_java
+    src/main/java/org/apache/arrow/gandiva/evaluator/ConfigurationBuilder.java
+    src/main/java/org/apache/arrow/gandiva/evaluator/JniWrapper.java
+    src/main/java/org/apache/arrow/gandiva/evaluator/ExpressionRegistryJniHelper.java
+    src/main/java/org/apache/arrow/gandiva/exceptions/GandivaException.java
+    GENERATE_NATIVE_HEADERS
+    gandivajni_headers
+    DESTINATION
+    ${JNI_HEADERS_DIR}/jni)
+  add_custom_target(gandiva_jni_headers)
```

Comment:
can we instead rename the gandivajni_headers target (in add_jar above) to gandiva_jni_headers ?

Reply:
we need both the targets to handle the java 7 case which means introducing a dummy target is simpler..instead of another conditional inside gandiva/jni/CMakeLists..